### ISSUE TITLE
rank_pct edge channel is identically zero on default config, including the default experiment recipe

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#114]
## Summary

On the repository's default configuration, the fourth multi-feature edge channel (`rank_pct`) is identically zero for every edge, and the documented default experiment recipe uses exactly that combination.

Found while rebuilding `AGENTS.md` and `docs/ARCHITECTURE.md` from current code (part of #101). No code change was made; this issue records the behavior.

## Mechanism

`configs/config.yaml` ships:

```yaml
  judge_value: 0.8             # Correlation threshold for edges (used when top_k == 0)
  top_k: 0                     # 0 = legacy global threshold; >0 = per-node top-K neighbours
  use_multi_feature_edges: true  # true => (E,4) edge_attr; false => scalar (E,)
```

`build_edges` populates `rank_pct` only on the top-K path:

```python
    if top_k > 0:
        rows, cols, kept_corr, rank_pct = _select_edges_topk(corr, top_k, top_k_metric)
    else:
        rows, cols, kept_corr = _select_edges_threshold(corr, judge_value)
        rank_pct = np.zeros_like(kept_corr)
```

and then stacks it regardless:

```python
        edge_attr = np.stack(
            [kept_corr, np.abs(kept_corr), kept_corr * kept_corr, rank_pct],
            axis=1,
```

So with `top_k=0` and `use_multi_feature_edges=true` — the shipped defaults — column 3 is a constant zero vector of length E.

## Why it matters

**The default experiment recipe hits this.** `docs/DEFAULT_EXPERIMENT_RECIPE.md`, referenced from `AGENTS.md`, specifies a *shuffled static threshold* graph together with multi-feature edges. Threshold mode means `top_k == 0`, so production-style confirmation runs train with a dead edge channel.

**The remaining channels are not independent either.** Columns 0, 1, and 2 are `corr`, `|corr|`, and `corr²` — all deterministic pointwise functions of a single scalar. They still give the model a fixed nonlinear basis over `corr`, which is not worthless, but combined with a constant column 3 it means the (E, 4) tensor carries the information content of one number per edge.

**`edge_feature_dim` still reports 4.** The model is sized for four channels and allocates parameters for the constant one.

**The docs asserted otherwise.** `AGENTS.md` described the four channels as `[corr, |corr|, corr^2, rank_pct]` with no indication that the fourth is inert outside top-K mode, which is presumably why this went unnoticed. PR #112 corrects the wording; this issue is about the behavior.

## What is not being claimed

A constant input channel is not a correctness bug. It does not corrupt results the way #107 or #111 can. The concern is that it is silent: nothing warns that a configured feature is inert, and the default recipe is the configuration that triggers it.

## Options worth weighing

- Compute a rank percentile in threshold mode too, so the channel means something under the default graph.
- Have `build_edges` emit three channels when `top_k == 0`, with `edge_feature_dim` agreeing, so no dead parameters are allocated.
- Warn at graph-build time when `use_multi_feature_edges` is enabled alongside `top_k=0`.
- Change the default recipe to top-K selection, if the dead channel turns out to be load-bearing for the current results.

Deciding this probably wants an ablation rather than a code review, since the honest question is whether the fourth channel was ever contributing.

## References

- `mci_gru/graph/correlation.py` (`build_edges`, `_select_edges_threshold`, `_select_edges_topk`)
- `mci_gru/graph/utils.py` (`edge_feature_dim`)
- `configs/config.yaml` (`graph.top_k`, `graph.use_multi_feature_edges`)
- `docs/DEFAULT_EXPERIMENT_RECIPE.md`
- #112 corrects the `AGENTS.md` description

**Comment by magilliam27** [comment]
## Investigation: the title understates this, and the real finding is that nothing guards it

Read-only investigation against `origin/main` @ `b416cbc`. No tracked file was edited. Provenance is labelled throughout: **[V-direct]** = I measured it in this session, **[V-agent]** = measured during the investigation run and structurally corroborated here, **[Inferred]**, **[Unknown]**.

## Bottom line

On the shipped default — `judge_value=0.8`, `top_k=0`, `use_multi_feature_edges=true`, which is also what `docs/DEFAULT_EXPERIMENT_RECIPE.md` pins — the `(E, 4)` edge tensor has **numerical rank 2**, and the two surviving directions are 0.9996-correlated. `rank_pct` being zero is one of two degenerate channels, not the whole story.

**But the dead channel is not the important part. It costs 261 parameters, 0.23% of the model, and is provably inert. The important part is that the shipped default combination has no test coverage through `build_edges` at all.**

## What the four channels actually contain

| channel | on the default config | status |
|---|---|---|
| 0 `corr` | +0.8194 … +0.8935 | the signal |
| 1 `\|corr\|` | **bit-identical to col 0** | exact duplicate |
| 2 `corr²` | pure function of col 0; **pearson(corr, corr²) = 0.9996** over the admitted band | near-collinear |
| 3 `rank_pct` | exactly `0.0` | dead |

**[V-direct]** Column 1 duplicating column 0 is **structural, not incidental**. `_select_edges_threshold` keeps `corr > judge_value`, and `mci_gru/config.py:351-352` rejects `judge_value <= 0`:

```
judge_value= -0.5 -> REJECTED (ValueError)
judge_value=  0.0 -> REJECTED (ValueError)
judge_value=  0.8 -> ACCEPTED
judge_value=  1.0 -> REJECTED (ValueError)
```

So the threshold path can only ever admit positive correlations, and `|corr| == corr` always. It cannot be configured otherwise.

**[V-direct]** The comparison discriminates. Same code, `judge_value=-0.5` forced past validation on a panel with genuine mid-range correlations: 130 negative edges kept, col 1 no longer identical to col 0, 4 distinct columns. Under `top_k=4`: `rank_pct` populated to 1.0. Both controls flip, so neither observation above is vacuous.

**[V-agent]** `corr` and `corr²` are 0.9996-correlated with condition number 87.5 over the narrow band the threshold admits. I initially described `corr²` as a legitimate nonlinear basis — that is true in principle and misleading in practice, because over `[0.80, 0.98]` it is very nearly affine in `corr`. Correcting my own earlier framing.

## The zero channel is bitwise inert

**[V-agent]** `GATConv.lin_edge` is constructed with `bias=False`, and `message()` never sees `edge_attr` — it enters only through attention. Consequences, each measured with a control:

- 261 parameters read column 3 (0.2269% of the default model).
- Their gradient is **exactly zero**; control with random values in column 3 gives non-zero gradient.
- Perturbing only those weights by +1000 leaves the output **bitwise identical**; control with real `rank_pct` changes it by 2.55.
- They still shrink under AdamW `weight_decay=1e-3`, consuming optimizer state while never affecting the loss.

A useful corollary: a constant *one* in column 3 **would** change the output, because LeakyReLU precedes the softmax. So "no ranking available" could not be signalled by a nonzero sentinel without changing behaviour — the charitable reading of the zero as a deliberate signal is ruled out mechanically.

## The actual defect: no coverage

**[V-agent]** A mutation that populates `rank_pct` with a real percentile in threshold mode **passes the full suite unchanged — 532 passed, 3 skipped, identical to baseline.** The control mutation, zeroing `rank_pct` in *top-K* mode, fails `TestMultiFeatureEdges::test_shape_and_columns`. So the harness can fail tests; nothing observes threshold-mode `rank_pct`.

**[V-direct]** Corroborated structurally without re-running the suite: every `GraphBuilder` constructed in `tests/` with `use_multi_feature_edges=True` passes `top_k` ∈ {1, 2, 3} — `test_dynamic_graph_updates.py:594,627,675,694` and `test_phase3_graph_and_walkforward.py:53`. `tests/test_inference_edge_dim.py` asserts the width is 4 but only calls `edge_feature_dim(GraphConfig(...))`; it never reaches `build_edges`, so it cannot see column content.

`TestMultiFeatureEdges::test_shape_and_columns` is the test that *looks* like it guards the four-column layout. It is built with `top_k=3`, so it does not.

## This was an unnoticed interaction, not a design decision

**[V-agent]** The roadmap designed multi-feature edges and top-K as a single lever — `graph_signal_upgrades_c28cf640.plan.md:9` and `:108` pair them. Threshold-mode multi-feature edges are never contemplated. Commit `837ee87` introduced `rank_pct = np.zeros_like(...)` while `use_multi_feature_edges` still defaulted to `false`, so the combination was opt-in. Commit `1ac15db` flipped that default to `true` and touched only that line — leaving `top_k=0` — with no mention of `rank_pct`. `docs/adr/` does not exist. No comments on this issue until now.

`AGENTS.md:155` and `docs/ARCHITECTURE.md:289-290` do state the channel is zero in threshold mode, but PR-112's own body frames that as **correcting a falsehood**, not recording a decision.

## Options

| | consequence |
|---|---|
| **A. Leave, document** | Zero risk, zero churn. `docs/FULL_FEATURE_FACTORIAL_ABLATION.md:42-45` still carries the text PR-112 corrected elsewhere and would need fixing. |
| **B. Warn at graph-build time** when `use_multi_feature_edges and top_k == 0` | Natural home is `mci_gru/graph/builder.py:128-135`, which already logs `n_feat`. No width change, no evidence break, one new test. Addresses the "it is silent" complaint at no risk. |
| **C. Narrow to 3 in threshold mode** | **[V-agent]** Breaks off-tree checkpoints — `load_state_dict` `RuntimeError` on a 4-wide checkpoint into a 3-wide model — and because `edge_feature_dim` is *derived, never persisted*, every previously-saved config silently re-derives to the new width. Nothing in-tree breaks: all committed configs predate the flags and resolve to 1. Test churn measured at exactly 1 test. Gain: 261 params. Also changes glorot `fan_in`, so surviving columns' init shifts and a narrowed run is **not** a controlled comparison against existing evidence. |
| **D. Populate `rank_pct` in threshold mode** | **Reject.** [V-agent] Out-degree varies wildly in threshold mode (measured min 0, max 19), so the channel would encode node degree as much as edge strength — an edge at `corr≈0.960` sits at `rank_pct=0.5` in a degree-2 row and near the top of a degree-19 row. Under top-K every row has exactly K, which is what makes it commensurate. Worse, the tensor shape is unchanged, so nothing would signal that pre- and post-change runs are different experiments. |
| **E. Switch the recipe to top-K** | **Contradicted by recorded evidence.** [V-agent] `docs/ABLATION_NOTEBOOK_RESULTS_REPORT_2026-04-30.md:48-49` — top-K "underperformed badly", `top_k=20` mean decision score −1.3704 vs threshold; `:179-180` concludes "Keep `graph.top_k=0`. Keep `graph.use_multi_feature_edges=true`." **That decision is precisely what locked in the dead channel.** |

### Recommendation (labelled as a recommendation)

**Add the missing test first, then take Option B. Gate Option C behind an ablation, and reject D and E.**

The parameter saving from narrowing is a rounding error against silently invalidating the width derivation for every off-tree checkpoint and making new runs uncontrolled against the whole recorded ablation corpus. That trade is not worth taking blind.

On the test, one refinement to how it should be written: a test authored against current behaviour would *pin* `col3 == 0` and make the eventual fix harder. Suggest splitting it — one test asserting the shape and the col 0/1/2 semantics on the **shipped default** combination, and a separate, deliberately-named `test_rank_pct_is_inert_in_threshold_mode` that documents the defect. When the defect is resolved, that second test is deleted rather than hunted for inside a larger assertion block.

## Found in passing, separate from this issue

1. **The venv's editable install resolves to the protected checkout.** `C:\Users\magil\MCI-GRU\.venv\...\__editable__.mci_gru-0.1.0.pth` points at `C:\Users\magil\MCI-GRU\mci_gru`, which sits on `codex/paper_trade_scrape @ e286649`. Any script run without an explicit `sys.path` insert silently exercises that stale tree. `scripts/run_pytest_isolated.py` is safe — it prepends `REPOSITORY_ROOT` at line 220. This nearly corrupted the investigation and is worth a line in `CLAUDE.md`.
2. **`CLAUDE.md` pins "40 dirty entries" for the protected checkout; the real count is 42**, unchanged before and after this session. The count is not an invariant and probably should not be stated as one.
3. **`edge_feature_dim` is never persisted** — only its inputs are. Recording the resolved value in run metadata would decouple this and make Option C far safer. Overlaps #144.
4. **`docs/FULL_FEATURE_FACTORIAL_ABLATION.md:42-45`** still asserts four populated channels — the exact text PR-112 corrected in `AGENTS.md` and `ARCHITECTURE.md` but missed here.
5. **[V-agent]** `|corr|` is redundant in top-K `metric=corr` mode too, most of the time, because selecting by signed correlation also tends to pick only positive edges. Only `metric=abs_corr` reliably makes column 1 independent. The redundancy is broader than threshold mode.
6. `paper_trade/scripts/compare_regime.py:248` reimplements the width rule as a hardcoded `4` rather than calling the shared helper, and already diverges for lead-lag configs. Noted only — `paper_trade/` is out of scope.

**Comment by magilliam27** [comment]
## The wider finding is filed as #157

While measuring for this issue, the graph itself turned out to be a bigger problem than its edge features.

`judge_value=0.8` sits **at or above the 99th percentile** of the correlation distribution on the 110-name production panel. On a January 2025 build date that leaves **36 edges among 206 names, with 84% of names isolated**; the median node has degree 0 in four of the five snapshots measured. PyG's `GATConv` defaults to `add_self_loops=True`, so nothing crashes — the GAT simply performs no cross-sectional aggregation for most names and reduces to a node-wise transform.

So this issue's degenerate edge features are riding on an adjacency that is close to empty. Both are worth knowing; #157 is the larger of the two.

PR-156 does not address #157 and is not blocked by it. It takes this issue's third listed option — warn at graph-build time — and adds the coverage whose absence let the dead channel ship.

Sequencing note recorded on #157: the graph-zeroed ablation it asks for is more informative once this issue's work lands, so the ablation is not measuring a degenerate feature tensor.
